experimental/packages/otlp-exporter-base/src/platform/browser/util.ts

```diff
@@ -59,49 +66,117 @@ export function sendWithXhr(
-  };
+    const defaultHeaders = {
+      'Accept': 'application/json',
+      'Content-Type': 'application/json',
+    };
 
-  Object.entries({
-    ...defaultHeaders,
-    ...headers,
-  }).forEach(([k, v]) => {
-    xhr.setRequestHeader(k, v);
-  });
+    Object.entries({
+      ...defaultHeaders,
+      ...headers,
+    }).forEach(([k, v]) => {
+      xhr.setRequestHeader(k, v);
+    });
 
-  xhr.send(body);
+    xhr.send(body);
 
-  xhr.onreadystatechange = () => {
-    if (xhr.readyState === XMLHttpRequest.DONE) {
-      if (xhr.status >= 200 && xhr.status <= 299) {
-        clearTimeout(exporterTimer);
-        diag.debug('xhr success', body);
-        onSuccess();
-      } else if (reqIsDestroyed) {
-        const error = new OTLPExporterError(
-          'Request Timeout', xhr.status
+    xhr.onreadystatechange = () => {
+      if (xhr.readyState === XMLHttpRequest.DONE && reqIsDestroyed === false) {
+        if (xhr.status >= 200 && xhr.status <= 299) {
+          diag.debug('xhr success', body);
+          onSuccess();
+          clearTimeout(exporterTimer);
+          clearTimeout(retryTimer);
+        } else if (xhr.status && isExportRetryable(xhr.status) && retries > 0) {
+          let retryTime: number;
+          minDelay = DEFAULT_EXPORT_BACKOFF_MULTIPLIER * minDelay;
+
+          // retry after interval specified in Retry-After header
+          if (xhr.getResponseHeader('Retry-After')) {
+            retryTime = retrieveThrottleTime(xhr.getResponseHeader('Retry-After')!);
+          } else {
+            // exponential backoff with jitter
+            retryTime = Math.round(Math.random() * (DEFAULT_EXPORT_MAX_BACKOFF - minDelay) + minDelay);
+          }
+
+          retryTimer = setTimeout(() => {
+            sendWithRetry(retries - 1, minDelay);
+          }, retryTime);
+        } else {
+          const error = new OTLPExporterError(
+            `Failed to export with XHR (status: ${xhr.status})`,
+            xhr.status
+          );
+          onError(error);
+          clearTimeout(exporterTimer);
+          clearTimeout(retryTimer);
+        }
+      }
+    };
+
+    xhr.onabort = () => {
+      if (reqIsDestroyed) {
+        const err = new OTLPExporterError(
+          'Request Timeout'
         );
-        onError(error);
-      } else {
-        const error = new OTLPExporterError(
-          `Failed to export with XHR (status: ${xhr.status})`,
-          xhr.status
+        onError(err);
+      }
+      clearTimeout(exporterTimer);
+      clearTimeout(retryTimer);
+    };
+
+    xhr.onerror = () => {
+      if (reqIsDestroyed) {
+        const err = new OTLPExporterError(
+          'Request Timeout'
         );
-        clearTimeout(exporterTimer);
-        onError(error);
+        onError(err);
       }
-    }
+      clearTimeout(exporterTimer);
+      clearTimeout(retryTimer);
+    };
   };
+
+  sendWithRetry();
+}
+
+function retrieveThrottleTime(retryAfter: string): number {
```

Comment:
This parameter type looks abstruse to me. The comment below suggests that the `retryAfter` can either be a Date object or an integer, but the explicit type here suggests it is a string. 

The call site of this function pass in a value from `xhr.getResponseHeader()`, which should be a string. This seems like diverging from the mocks in the test.

Reply:
@legendecas 

Well the `Retry-After` can be a a Date or an integer (representing seconds) and when I retrieve the Retry-After with `getResponseHeader()`, it returns a string so `retrieveThrottleTime` parameter is set to string type. But I'm able to use `typeof` inside `retrieveThrottleTime ` to check if `retryAfter` argument is an object or not. 

Maybe I can change `function retrieveThrottleTime(retryAfter: string)` to be `function retrieveThrottleTime(retryAfter: string | object)`. Do you have any recommendations on how else I can implement this logic? Appreciate the help.

Reply:
What do you think about https://github.com/open-telemetry/opentelemetry-js/pull/3577/files#diff-67f1efb4b6dca82d1da8bd9d718dfca1b9126597c8e5b225a50f3926c9874d35R72?

Reply:
This works when the retry-after is in seconds but it doesn't work when retry-after is a date. When the retry-after is a date, the `retrieveThrottleTime` won't work because `seconds` will be `NaN`.

Example:

```js
function retrieveThrottleTime(retryAfter?: string | null): number {
    if (retryAfter == null) {
      return -1;
    }

    console.log("retry after is ", retryAfter)
    // logs => Wed Feb 08 2023 11:53:14 GMT-0600 (Central Standard Time)

    const seconds = Number.parseInt(retryAfter, 10);
    console.log('seconds are ', seconds)
   // logs => NaN

    if (Number.isInteger(seconds)) {
      return seconds > 0 ? seconds * 1000 : -1;
    }
    // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After#directives
    const delay = new Date(retryAfter).getTime() - Date.now();
    if (delay >= 0) {
      return delay * 1000;
    }
    return 0;
}
```

Reply:
What about this small tweak to address the above issue:

```js
function retrieveThrottleTime(retryAfter?: string | null): number {
    if (retryAfter == null) {
      return -1;
    }

    const seconds = Number.parseInt(retryAfter, 10);

    if (isNaN(seconds)) {
      // https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After#directives
      const delay = (new Date(retryAfter).getTime() - Date.now()) / 1000;
      if (delay >= 0) {
        return delay * 1000;
      }
      return 0;
    } else {
      return seconds > 0 ? seconds * 1000 : -1;
    }
}
```

Reply:
Do you referring to this test case? https://github.com/open-telemetry/opentelemetry-js/pull/3577/files#diff-8baa107c057bf551118789c29ad093d016a7ddb1a26a9c5ad02d67ff386e1dffR63

IIUC `Number.isInteger(NaN)` returns false so when `retry-after` response header is a date string, it would be parsed with `new Date` as suggested by the MDN doc.

Reply:
Yes I'm referring to that test case. Thanks for the clarification on this:

> IIUC Number.isInteger(NaN) returns false so when retry-after response header is a date string, it would be parsed with new Date as suggested by the MDN doc.

I have a question on this part of the function:

```js
const delay = new Date(retryAfter).getTime() - Date.now();
    if (delay >= 0) {
      return delay * 1000;
    }
```

Shouldn't that be:

```js
const delay = new Date(retryAfter).getTime() - Date.now();
    if (delay >= 0) {
      return delay;
    }
```

Otherwise if we do `return delay * 1000` then the `delay` would be too large. For example, if the retry-after date is `new Date('2023-01-20T00:00:01.000Z')` and the now date is `new Date('2023-01-20T00:00:00.000Z')`, then this function will return the delay of `1000000` milliseconds which is 100 seconds but we need 1000 milliseconds for a 1 second delay for this test case.

Reply:
Ah, yes, you're right. Thanks for pointing this out.

Reply:
I've updated this function to match your feedback.